torch/csrc/distributed/c10d/init.cpp

```diff
@@ -734,21 +734,25 @@ They are used in specifying strategies for reduction collectives, e.g.,
                 >>> ddp_model._register_comm_hook(state = None, hook = allreduce)
 
             .. warning ::
-                ``get_future`` API supports only NCCL backend. The ``torch._C.Future`` object
-                returned by this API can be used in ``DistributedDataParallel._register_comm_hook``,
-                but it is subject to some subtle differences compared to ``torch.futures.Future``
-                due to compromises made for performance reasons.
+                ``get_future`` API supports only NCCL backend and single-process single-device mode.
+                The ``torch._C.Future`` object returned by this API can be used in
+                ``DistributedDataParallel._register_comm_hook``, but it is subject to some subtle
+                differences compared to ``torch.futures.Future`` due to compromises made for performance
+                reasons.
 
                 In the example above, ``allreduce`` work will be done on GPU using NCCL backend,
                 ``fut.wait()`` will return after synchronizing the appropriate NCCL streams
                 with PyTorch's default device streams to ensure we can have asynchronous CUDA
-                execution and it does not wait for the entire operation to complete on GPU.
-                If ``NCCL_BLOCKING_WAIT`` is enabled, in that case, it would wait for the entire
-                operation to complete before returning. In addition, if a callback function was
-                added by ``fut.then()``, it will wait until WorkNCCL's wait returns and invoke
-                the callback inline.
-
-                Note that ``fut.done()`` returns if the work was completed on the GPU.
+                execution and it does not wait for the entire operation to complete on GPU. Note that
+                ``FutureNCCL``  does not support ``NCCL_BLOCKING_WAIT`` flag or NCCL's ``barrier()``.
+                In addition, if a callback function was added by ``fut.then()``, it will wait until
+                ``WorkNCCL``'s NCCL streams synchronize with a new stream from device's stream pool and
+                invoke the callback inline after running the callback on the new stream. ``fut.then()``
+                will return another ``FutureNCCL`` that holds the return value of the callback and the
+                stream that runs the callback.
+
+                Note that ``fut.done()`` returns if work's NCCL streams were synchronized with PyTorch's
+                default device streams.
```

Comment:
This doesn't seem to be right? In `completed()` we actually check if the work has completed on the GPU using `cudaEventQuery`.